feat(FileUpload) Add new beta File Upload component

mturley opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Closes #3735 

Preview: https://patternfly-react-pr-3865.surge.sh/documentation/react/components/fileupload

<!-- Are there any upstream issues or separate issues you need to reference? -->
**Additional issues**: Based on core implementation from https://github.com/patternfly/patternfly-next/pull/2681, https://github.com/patternfly/patternfly-next/pull/2731 and https://github.com/patternfly/patternfly-next/pull/2776

patternfly-build commented:
PF4 preview: https://patternfly-react-pr-3865.surge.sh

mturley commented:
@tlabaj I still have a few small TODO items here, but it people are free I could use some early reviews since we want this merged by Friday. All that's left to do is add the loading spinner case and tests, the component and docs are mostly done.

evwilkin commented:
@mturley this looks good to me so far, pending your remaining updates - thanks for your work on this!

seanforyou23 commented on packages/react-core/src/components/FileUpload/FileUploadField.tsx:
```diff
@@ -0,0 +1,163 @@
+  /** Flag to show if a file is being dragged over the field */
+  isDragActive?: boolean;
+  /** A reference object to attach to the FileUploadField container element. */
+  containerRef?: React.Ref<any>;
```
Are we able to type this? `HTMLDivElement`?

seanforyou23 commented on packages/react-core/src/components/FileUpload/FileUploadField.tsx:
```diff
@@ -0,0 +1,163 @@
+  id,
+  value = '',
+  filename = '',
+  onChange = (): any => undefined,
```
Are these default functions necessary? If not can we remove them, and if they are - can we type them?

seanforyou23 commented on packages/react-core/src/components/FileUpload/FileUploadField.tsx:
```diff
@@ -0,0 +1,163 @@
+  browseButtonText = 'Browse...',
+  clearButtonText = 'Clear',
+  clearButtonDisabled = !filename && !value,
+  containerRef = null as React.Ref<any>,
```
Are we able to type this? `HTMLDivElement`?

seanforyou23 commented:
Would using a [`FileReader`](https://developer.mozilla.org/en-US/docs/Web/API/FileReader) help with the blocking-ui behavior?

mturley commented:
Thanks for the review comments @seanforyou23. I'll fix up the types as you suggested.

As far as blocking file types that are not supported, I was thinking I would leave that up to the consumer (via dropzoneProps, see my second example), but maybe there's a sensible default I can set there. I didn't want to make an assumption here since the react-dropzone docs point out that MIME type determination is not reliable across platforms, and I don't see an easy way to reliably identify if a file is binary or not via FileReader or the File object.

https://react-dropzone.js.org/#!/Dropzone

> accept -- Set accepted file types. See https://github.com/okonet/attr-accept for more information. Keep in mind that mime type determination is not reliable across platforms. CSV files, for example, are reported as text/plain under macOS but as application/vnd.ms-excel under Windows. In some cases there might not be a mime type set at all. See: https://github.com/react-dropzone/react-dropzone/issues/276

I'm still not sure how to solve the blocking-ui behavior, I'm already using a FileReader. I'll keep looking into that for now, but if I don't find a solution today I wonder if it would be acceptable to merge the initial version of this component with this issue and address it in a follow-up. For most small text files, the load happens instantly.

seanforyou23 commented on packages/react-core/src/components/FileUpload/examples/FileUpload.md:
````diff
@@ -0,0 +1,151 @@
+
+The basic `FileUpload` component can handle simple text files via browse or drag-and-drop, loading them into memory and passing their contents as a string to an `onChange` prop.
+
+```js title=Simple-text-file isBeta
````
I think the `isBeta` flag on the individual examples is only needed if the main component/feature is not beta. Since the entire FileUpload is currently beta, we can probably omit these in favor of the main beta flag at the top.

mturley commented on packages/react-core/src/components/FileUpload/FileUpload.tsx:
```diff
@@ -0,0 +1,128 @@
+  id,
+  value = '',
+  filename = '',
+  onChange = (): any => undefined,
```
As discussed, just noops for optional callbacks so I don't have to check that they are defined on each call. I don't think the typing is necessary on them at all, just replaced them all with `() => {}`. Let me know if you think we should use something better there.

mturley commented:
Ok @seanforyou23 I'm happy with this as it is now, we just need some tests and some more reviews! Feel free to jump in if you start before I do tomorrow. I'll be offline between about 1:30-3:30 and then I'll be around for the last hour or two before code freeze.

mturley commented:
One thing I do want to note is that the "file type and size summary" I added to support non-text files is something I made up in response to feedback, and we will probably need to go back through the design and core implementation process to support a better-looking version of that.

christiemolloy commented:
I think we should show an editable and non-editable version of this, and label them

<img width="647" alt="Screen Shot 2020-03-06 at 2 32 06 PM" src="https://user-images.githubusercontent.com/20118816/76116113-6799ad80-5fb7-11ea-9690-96ab63c1bf62.png">

Do you this focus should just go to the file upload component and not the text? Would probably be a change in Core, but wondering what people think

<img width="652" alt="Screen Shot 2020-03-06 at 2 34 19 PM" src="https://user-images.githubusercontent.com/20118816/76116244-a62f6800-5fb7-11ea-83df-d0898b6a7d01.png">

What is the purpose of adding children? When I check the children box , the border is cut off by the child
<img width="637" alt="Screen Shot 2020-03-06 at 2 35 53 PM" src="https://user-images.githubusercontent.com/20118816/76116373-e0990500-5fb7-11ea-9d46-559b625be219.png">

is there a way to restrict this so that if someone chooses loading they have to select showPreview?

<img width="654" alt="Screen Shot 2020-03-06 at 2 37 24 PM" src="https://user-images.githubusercontent.com/20118816/76116507-1e962900-5fb8-11ea-9214-0d0c1622406a.png">

seanforyou23 commented:
The recommended a11y practice is to send focus back to whatever element had it before the file selector process started. I think this would mean the button in most cases, but it does make me think about the case of dragging onto the dropzone area. Even though the component container real estate is the droppable area, it doesn't receive focus (nor does it need to) so I'd still think to send focus to the browse button after the drop operation also. It seems to be working as intended when you use the browser button, but after dropping currently focus seems to be lost to the page.

@christiemolloy I think `children` was intended for consumers to be able to have a custom preview area if they wanted some special formatting. Not sure about the border thingie.

mturley commented:
@christiemolloy I think maybe I should remove that last example if it's causing confusion. Also I changed the `showPreview` prop to a `hideDefaultPreview` which defaults to false in my last commit, so that might help with the confusion there a little.

children is intended to be used only if hideDefaultPreview is also set, since they are for a custom preview. That border will not be cut off if it's used this way. So maybe if children are passed, the preview should be automatically hidden. (but it's still useful to be able to hide the preview entirely).

mturley commented:
Per discussion with Titani, I removed the default preview for a non-text file since it needs design and core consideration. Here's what it looked like for posterity, after this is merged I'll open another WIP PR to play around with it:

<img width="762" alt="Screenshot 2020-03-06 13 52 13" src="https://user-images.githubusercontent.com/811963/76130273-409fa380-5fd8-11ea-948e-ba7deb9741c3.png">

mturley commented:
Thanks to @seanforyou23 for helping me out here with tests and helping me with some decisions!

mcoker commented:
@mturley this looks great!

I'm curious why we need the `tabindex` on the outer `.pf-c-file-upload` element?

Also I think you're on an older version of core? I'm not seeing the blue border on drag/hover.

I wonder if we should limit the drag/hover styles to the form elements, and not apply them over any messaging underneath the form elements. wdty @mcarrano? 

<img width="787" alt="Screen Shot 2020-03-06 at 5 30 10 PM" src="https://user-images.githubusercontent.com/35148959/76130483-85bfd780-5fd0-11ea-95af-052e1069cf80.png">

FWIW, I used a 5mb jpg image to upload to the "simple text file" example, and my browser froze and was unresponsive for 3 minutes before updating the file preview. Not sure if there's anything we could do about that.

mturley commented:
@christiemolloy:

> I think we should show an editable and non-editable version of this, and label them

@mcarrano:

> The only thing I was not sure about was whether there was a way to optionally edit the text uploaded into the text area. Looks like that is currently read-only in the example, but if I recall our conversation, we wanted to make editing optional.

I can add a prop like `allowEditingUploadedFile` and show an example with that, would that meet both your needs? Sorry I overlooked that.

@mcoker:

> I'm curious why we need the `tabindex` on the outer `.pf-c-file-upload` element?

I was just duplicating what I saw in core, I'm not sure exactly 😆 but I can do whatever we need there.

> Also I think you're on an older version of core? I'm not seeing the blue border on drag/hover.

You're right! I need to bump that version, sorry.

> FWIW, I used a 5mb jpg image to upload to the "simple text file" example, and my browser froze and was unresponsive for 3 minutes before updating the file preview. Not sure if there's anything we could do about that.

Yeah, that's unfortunately what happens when you take that much data and shove it into a JS variable in React state. We could maybe set a sensible default for the maximum size allowed for that basic text file case, and advise the consumer not to use `type="text"` if they need larger files. If you don't use `type="text"`, it doesn't try to read the file at all, it just passes a handle to the consumer so they can do what they need to to read the file themselves. See my "custom file preview" example, you can drop a large file there and it'll just show the size without doing anything else.

mcoker commented:
> I was just duplicating what I saw in core, I'm not sure exactly 😆 but I can do whatever we need there.

@mturley hmm, I'm looking on [master](https://pf4.patternfly.org/documentation/core/components/fileupload) and don't see `tabindex` defined on any of the file upload component elements. Maybe it was there at some point and removed?

mturley commented:
@mcoker that's probably the case. I'll bump the core version and look over it all to make sure I'm up to date.

seanforyou23 commented:
I think generally if `tabindex` is present we need to also ensure there's an accessible name for that element. Otherwise, as it stands currently the extra ways we can interact with the container don't get exposed to consumers using AT, and we have a mismatch between visual, non-visual, and keyboard experience.

That said, having two focusable elements to facilitate the same operation (for **each** file upload component on the page) is redundant and exposing this in an intuitive way to SRU will be unnecessarily involved without any gain, but definitely more clutter. The names need to be unique, and there are already many elements to wade through on any given web page these days so if we can have less nodes that are better described it will be better than having overwhelming number of options with names close enough to be confusing. Not worth the extra cognitive load it brings. 

We could either: Add an aria-label to the container that has tabindex or, we could remove tabindex. My vote is to go with the latter and only use the container for drag/drop purposes - then nail making the component browse/clear btn names unique among many of a same page. That would make for the best accessible experience in my opinion. I hope it helps, lemme know what you think!

mturley commented:
@mcoker I'm not actually seeing a `tabindex` in my source, where are you seeing that?

mturley commented:
Ah, I think the tabindex is being applied automatically by the library I'm using for drag-and-drop (react-dropzone). I can override that and prevent it

mcarrano commented:
> I wonder if we should limit the drag/hover styles to the form elements, and not apply them over any messaging underneath the form elements. wdty @mcarrano?

Good catch @mcoker . I don't think this is a big deal, but if we can restrict the styles to form elements, that would be preferrable.

mturley commented:
@mcarrano @mcoker In the core examples, the `.pf-c-file-upload__message` element is inside the `.pf-c-file-upload` container which gets the `.pf-m-drag-hover` class applied to it. For the message to be outside the drag/hover box, we'll need to change that in core.

The other option is, we can remove the styles for error/info messaging from FileUpload entirely, since the user can get the same result by wrapping the whole thing in a FormGroup component with its own `helperText`/`helperTextInvalid` props. Maybe that's what the examples should use instead?

mcoker commented:
> remove the styles for error/info messaging from FileUpload entirely, since the user can get the same result by wrapping the whole thing in a FormGroup component with its own helperText/helperTextInvalid props.

That sounds great to me if you think that makes sense from the developer's perspective.

mturley commented:
@seanforyou23 :

> We could either: Add an aria-label to the container that has tabindex or, we could remove tabindex. My vote is to go with the latter and only use the container for drag/drop purposes - then nail making the component browse/clear btn names unique among many of a same page.

I agree. I'm omitting the `tabindex` from react-dropzone's `getRootProps()` there. 
For making the browse/clear button names unique, what specific attributes do you think we need there? Based on the core implementation, I gave the filename field a unique id and aria-label, and set its aria-describedby to the Browse button which has a unique id (all generated based on the `id` prop passed by the consumer for the whole component). But the Clear button doesn't have its own unique id. Do we need to change that (both here and in core) or is it enough to have it be adjacent to the uniquely-identified filename stuff?

mturley commented:
@mcoker I think it does, it'll match how the developer needs to provide validation messages on other types of form fields. I'll make that change in my examples and remove the `message` prop, and we'll have to loop back and remove the error message HTML/CSS from the core examples and maybe use formgroup stuff there too. (cc @christiemolloy )

mturley commented:
@mcoker here's what it looks like using a FormGroup for the message instead. The dropzone area is now restricted to the form fields:

<img width="1000" alt="Screenshot 2020-03-09 11 32 30" src="https://user-images.githubusercontent.com/811963/76230240-d108fe80-61f9-11ea-8b67-6f76f0f6075b.png">

It is worth noting that if the consumer passes `hideDefaultPreview` and `children` to render their own custom preview (e.g. an image thumbnail if they are handling images), that will still be within the dropzone (so they could drag/drop another image on top of the previous image). If they don't have any preview at all (`hideDefaultPreview` and no `children`) then the dropzone will be only the filename/browse/clear bar.

mturley commented:
Ok, I just rebased and resolved merge conflicts caused by my version bump to `@patternfly/patternfly@2.68.3`. I think I've addressed all the feedback so far, this should be ready to go!

mturley commented:
Yeah, 51 commits, I guess I should squash 😆 if anyone wants to back and look at the original history I'll leave it preserved here: https://github.com/patternfly/patternfly-react/compare/2aac12f9683d44cfa603236f68adaee8b082fcbf...mturley:3735-file-upload-component-presquash

mturley commented:
Left the pf version bump unsquashed in case we need to revert/replace it, since it was the thing causing conflicts.

@seanforyou23 @mcoker would you mind re-approving?

patternfly-build commented:
Your changes have been released in: 
 - @patternfly/react-catalog-view-extension@1.4.27
 - @patternfly/react-charts@5.3.12
 - @patternfly/react-core@3.145.2
 - @patternfly/react-docs@4.20.33
 - @patternfly/react-icons@3.15.11
 - @patternfly/react-inline-edit-extension@2.17.27
 - demo-app-ts@3.28.2
 - @patternfly/react-styled-system@3.8.10
 - @patternfly/react-styles@3.7.8
 - @patternfly/react-table@2.28.8
 - @patternfly/react-tokens@2.8.8
 - @patternfly/react-topology@2.14.27
 - @patternfly/react-virtualized-extension@1.4.28

Thanks for your contribution! :tada:
